src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,392 @@
+            if (ifStatement == null)
+            {
+                return;
+            }
+
+            if (ifStatement.OverlapsHiddenPosition(cancellationToken))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(
+                new MyCodeAction(
+                    GetInvertIfText(),
+                    c => InvertIfAsync(document, ifStatement, c)));
+        }
+
+        private async Task<Document> InvertIfAsync(
+            Document document, 
+            SyntaxNode ifStatement, 
+            CancellationToken cancellationToken)
+        {
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            return document.WithSyntaxRoot(
+                // this returns the root because VB requires changing context around if statement
+                GetRootWithInvertIfStatement(
+                    document, model, ifStatement, cancellationToken));
+        }
+
+        internal SyntaxNode Negate(
+    SyntaxNode expression,
+    SyntaxGenerator generator,
+    ISyntaxFactsService syntaxFacts,
+    SemanticModel semanticModel,
+    CancellationToken cancellationToken)
+        {
+
+            if (syntaxFacts.IsParenthesizedExpression(expression))
+            {
+                return syntaxFacts.Parenthesize(
+                    Negate(
+                        syntaxFacts.GetExpressionOfParenthesizedExpression(expression),
+                        generator,
+                        syntaxFacts,
+                        semanticModel,
+                        cancellationToken))
+                    .WithLeadingTrivia(expression.GetLeadingTrivia())
+                    .WithTrailingTrivia(expression.GetTrailingTrivia());
+            }
+            if (syntaxFacts.IsBinaryExpression(expression))
+            {
+                return GetNegationOfBinaryExpression(expression, generator, syntaxFacts, semanticModel, cancellationToken);
+            }
+            else if (syntaxFacts.IsLiteralExpression(expression))
+            {
+                return GetNegationOfLiteralExpression(expression, generator, semanticModel);
+            }
+            else if (syntaxFacts.IsLogicalNotExpression(expression))
+            {
+                return (GetNegationOfLogicalNotExpression(expression, generator, syntaxFacts, semanticModel));
+            }
+            return generator.LogicalNotExpression(expression);
+        }
+
+        private SyntaxNode GetNegationOfBinaryExpression(
+            SyntaxNode expressionNode,
+            SyntaxGenerator generator,
+            ISyntaxFactsService syntaxFacts,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken)
+        {
+            syntaxFacts.GetPartsOfBinaryExpression(expressionNode, out var leftOperand, out var rightOperand);
+            var operatorToken = syntaxFacts.GetOperatorTokenOfBinaryExpression(expressionNode);
```

Comment:
nit: just have GetPartsOfBinaryExpression also include the token.  Then add a GetPartsOfBinaryExpression extension method to ISyntaxFactsExtensions that keeps hte old overload and forwards it to the new method.